### PR TITLE
Add conditional rendering example

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a new example for the `VK_EXT_conditional_rendering` extension. With this extension one can conditionally disable draw and dispatch commands based on a dedicated buffer. The sample render a complex glTF sample and uses such a conditional buffer to toggle visibility for the glTF's scene nodes based on user input.

The sample also includes a tutorial in the readme.

**Important note**: I had to make a slight change to remove a conjugation of the rotation from one of the frameworks components (see [here](https://github.com/KhronosGroup/Vulkan-Samples/compare/master...SaschaWillems:conditional_render?expand=1#diff-4746ae3da524d8cc63f4e454636e84d0a00cd31b3f3111a941da474c8d91d367L87)). Initially, with this conjugation in place, the new glTF model did not display properly. Removing it fixed this and I tested all other samples to make sure that this didn't break any of the other glTF's used in this repo.

**Important note**: This sample requires new assets, so https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/17 needs to be merged first.

Sample was tested on Windows 10 and an NVIDIA RTX 2060.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/extensions/conditional_rendering/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+get_filename_component(FOLDER_NAME ${CMAKE_CURRENT_LIST_DIR} NAME)
+get_filename_component(PARENT_DIR ${CMAKE_CURRENT_LIST_DIR} PATH)
+get_filename_component(CATEGORY_NAME ${PARENT_DIR} NAME)
+
+add_sample(
+    ID ${FOLDER_NAME}
+    CATEGORY ${CATEGORY_NAME}
+    AUTHOR "Sascha Willems"
+    NAME "VK_EXT_conditional_rendering"
+    DESCRIPTION "todo"
```
Comment (TomAtkinsonArm) [inline samples/extensions/conditional_rendering/CMakeLists.txt]: We can copy and paste the description from the samples readme here?

**Review by TomAtkinsonArm — CHANGES_REQUESTED** [review]
Small change requested

Other than this it LGTM!! Its pretty fun to toggle different meshes on and off

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+	render_pass_begin_info.renderArea.offset.x   = 0;
+	render_pass_begin_info.renderArea.offset.y   = 0;
+	render_pass_begin_info.clearValueCount       = 2;
+	render_pass_begin_info.pClearValues          = clear_values;
+
+	for (int32_t i = 0; i < draw_cmd_buffers.size(); ++i)
+	{
+		VK_CHECK(vkBeginCommandBuffer(draw_cmd_buffers[i], &command_buffer_begin_info));
+
+		VkClearValue clear_values[2];
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp]: No need to have that clear_values inside the loop again, can use that one from above.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+		clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+		clear_values[1].depthStencil = {0.0f, 0};
+
+		VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp]: No need to have a new render_pass_begin_info here, can use that one from above. The only part that changes per loop iteration is the framebuffers[i].

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+		VkRect2D scissor = vkb::initializers::rect2D(width, height, 0, 0);
+		vkCmdSetScissor(draw_cmd_buffers[i], 0, 1, &scissor);
+
+		vkCmdBindPipeline(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline);
+		vkCmdBindDescriptorSets(draw_cmd_buffers[i], VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layout, 0, 1, &descriptor_set, 0, nullptr);
+
+		uint32_t idx = 0;
+		for (auto & node : linear_nodes)
+		{
+			glm::mat4 node_transform = node.node->get_transform().get_world_matrix();
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp]: I would move all those local variables (node_transform, offsets, ..., mat) closer to where they're used below, maybe not even needing a local variable.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+	update_uniform_buffers();
+}
+
+void ConditionalRendering::update_uniform_buffers()
+{
+	uniform_data.projection = camera.matrices.perspective;
+	uniform_data.view       = camera.matrices.view * glm::mat4(1.0f);
+	// @todo: comment
+	uniform_data.view       = glm::scale(uniform_data.view, glm::vec3(0.1f, -0.1f, 0.1f));
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp]: Why do you multiply camera.matrices.view with the identity, and scale it in a separate step?

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+void ConditionalRendering::on_update_ui_overlay(vkb::Drawer &drawer)
+{
+	if (drawer.header("Visibility"))
+	{
+		if (drawer.button("All"))
+		{
+			for (auto i = 0; i < conditional_visibility_list.size(); i++)
+			{
+				conditional_visibility_list[i] = 1;
+			}
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp]: Use std::fill() instead?

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+			update_visibility_buffer();
+		}
+		ImGui::SameLine();
+		if (drawer.button("None"))
+		{
+			for (auto i = 0; i < conditional_visibility_list.size(); i++)
+			{
+				conditional_visibility_list[i] = 0;
+			}
+			update_visibility_buffer();
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp:389]: Use std::fill() instead?

File: samples/extensions/conditional_rendering/conditional_rendering.cpp [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+	add_device_extension(VK_EXT_CONDITIONAL_RENDERING_EXTENSION_NAME);
+}
+
+ConditionalRendering::~ConditionalRendering()
+{
+	if (device)
+	{
+		vkDestroyPipeline(get_device().get_handle(), pipeline, nullptr);
+		vkDestroyPipelineLayout(get_device().get_handle(), pipeline_layout, nullptr);
+		vkDestroyDescriptorSetLayout(get_device().get_handle(), descriptor_set_layout, nullptr);
```
Comment (asuessenbach) [inline samples/extensions/conditional_rendering/conditional_rendering.cpp:41]: Where is the descriptor_set freed?

File: shaders/conditional_rendering/model.vert [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+} push_constants;
+
+layout (location = 0) out vec3 outNormal;
+layout (location = 1) out vec3 outColor;
+layout (location = 2) out vec3 outViewVec;
+layout (location = 3) out vec3 outLightVec;
+
+void main() 
+{
+	outNormal = inNormal;
```
Comment (asuessenbach) [inline shaders/conditional_rendering/model.vert]: No need to set outNormal here.

File: samples/extensions/conditional_rendering/CMakeLists.txt (hunk 1 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/CMakeLists.txt]: Good catch. Should do a search for "todo" in the future before submitting new PRs ^^

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 2 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp]: Good catch. Dunno why I duplicated this.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 3 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp]: Same as above. Thanks for noticing.

File: shaders/conditional_rendering/model.vert (hunk 9 above)
Comment (SaschaWillems) [reply shaders/conditional_rendering/model.vert]: Good catch. That line is indeed not necessary at all.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 4 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp]: Good point. Will change that.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 6 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp]: Good idea. Will use std:fill in here.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 7 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp:389]: Same as above, will move to std::fill.

File: samples/extensions/conditional_rendering/conditional_rendering.cpp (hunk 5 above)
Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/conditional_rendering.cpp]: Good question. Will combine the two steps.

**Review by asuessenbach — APPROVED** [review]

**Review by gpx1000 — APPROVED** [review]
LGTM

**Review by gary-sweet — COMMENTED** [review]
Doesn't break anything for me, and correctly reports that the extension is not supported.
Can't really add an approval given that I can't test the actual code though.

File: samples/extensions/conditional_rendering/CMakeLists.txt (hunk 1 above)
Comment (marty-johnson59) [reply samples/extensions/conditional_rendering/CMakeLists.txt]: Is this resolved now?  are we good to merge?

Comment (SaschaWillems) [reply samples/extensions/conditional_rendering/CMakeLists.txt]: Yes, I have since then added a proper description, so this has been resolved.

Comment (marty-johnson59) [reply samples/extensions/conditional_rendering/CMakeLists.txt]: OK, sounds good.  Then should be safe to assume we have Tom's approval as well - I'll merge.  Thanks!

**Pull request merged**